src/Microsoft.Diagnostics.Runtime/ClrValue.cs

```diff
@@ -0,0 +1,1077 @@
+            get
+            {
+                return _obj;
+            }
+        }
+
+        public override int Size
+        {
+            get
+            {
+                return _type.IsObjectReference ? (int)_type.GetSize(_obj) : _type.BaseSize;
+            }
+        }
+
+        public override ulong Address
+        {
+            get
+            {
+                return _address;
+            }
+        }
+
+        public override bool Interior
+        {
+            get
+            {
+                return _interior;
+            }
+        }
+
+        public override ClrElementType ElementType
+        {
+            get
+            {
+                return _type.ElementType;
+            }
+        }
+
+        public ClrValueImpl(ClrRuntime runtime, ulong address, ClrInstanceField field)
+            : base(runtime)
+        {
+            _address = address;
+            _type = field.Type;
+            _interior = field.IsValueClass;
+
+            if (_type.IsObjectReference)
+            {
+                var heap = _type.Heap;
+                if (!heap.ReadPointer(address, out _obj))
+                    throw new MemoryReadException(address);
+
+                _type = heap.GetObjectType(_obj);
+            }
+        }
+
+        public ClrValueImpl(ClrHeap heap)
+            : base(heap.Runtime)
+        {
+            _address = 0;
+            _type = heap.NullType;
+        }
+
+        public override ClrObject AsObject()
+        {
+            if (Interior)
+                throw new InvalidOperationException("Value is an interior pointer, not an object.");
+
+            return new ClrObject(_obj, _type);
+        }
+
+        public override ClrType Type
+        {
+            get
+            {
+                return _type;
+            }
+        }
+
+        protected override ClrType GetStructElementType()
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal class CorDebugValue : ClrValue
+    {
+        ICorDebug.ICorDebugValue _value;
+        int? _size;
+        ulong? _address;
+        ulong? _object;
+        ClrElementType _elementType;
+
+        protected override ulong Object
+        {
+            get
+            {
+                if (!_object.HasValue)
+                {
+                    ulong obj;
+                    if (!Runtime.ReadPointer(Address, out obj))
+                        throw new MemoryReadException(Address);
+
+                    _object = obj;
+                }
+
+                return _object.Value;
+            }
+        }
+
+        public override int Size
+        {
+            get
+            {
+                if (!_size.HasValue)
+                {
+                    uint size;
+                    _value.GetSize(out size);
+                    _size = (int)size;
+                }
+
+                return _size.Value;
+            }
+        }
+
+        public override bool Interior
+        {
```

Comment:
I haven't looked enough at the surrounding code, but this seemed a surprising implementation